src/main/java/com/rackspace/ceres/app/services/DelayedTrackingService.java

```diff
@@ -0,0 +1,145 @@
+
+import com.rackspace.ceres.app.config.AppProperties;
+import com.rackspace.ceres.app.config.DownsampleProperties;
+import com.rackspace.ceres.app.model.PendingDownsampleSet;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Profile;
+import org.springframework.data.redis.core.ReactiveStringRedisTemplate;
+import org.springframework.data.redis.core.script.RedisScript;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.time.Instant;
+import java.util.List;
+
+import static com.rackspace.ceres.app.utils.DateTimeUtils.*;
+
+@Service
+@Slf4j
+@Profile("downsample")
+public class DelayedTrackingService {
+  private final DownsampleProperties properties;
+  private final AppProperties appProperties;
+  private final ReactiveStringRedisTemplate redisTemplate;
+  private final RedisScript<String> redisGetDelayedJob;
+
+  @Autowired
+  public DelayedTrackingService(ReactiveStringRedisTemplate redisTemplate,
+                                DownsampleProperties properties,
+                                AppProperties appProperties,
+                                RedisScript<String> redisGetDelayedJob) {
+    this.redisTemplate = redisTemplate;
+    this.redisGetDelayedJob = redisGetDelayedJob;
+    this.properties = properties;
+    this.appProperties = appProperties;
+  }
+
+  public Flux<String> claimJob(Integer partition) {
+    String hostName = System.getenv("HOSTNAME");
+    long now = Instant.now().getEpochSecond();
+    long maxJobDuration = properties.getMaxDownsampleJobDuration().getSeconds();
+    return redisTemplate.execute(
+        this.redisGetDelayedJob,
+        List.of(),
+        List.of(partition.toString(),
+            hostName == null ? "localhost" : hostName,
+            Long.toString(now),
+            Long.toString(maxJobDuration))
+    );
+  }
+
+  public Mono<?> freeJob(int partition) {
+    log.trace("free job partition {}", partition);
+    return redisTemplate.opsForValue().set(encodeJobKey(partition), "free");
+  }
+
+  public Flux<String> getDelayedTimeSlots(Integer partition, String group) {
+    String key = encodeDelayedTimeslotKey(partition, group);
+    return this.redisTemplate.opsForSet().scan(key)
+        .sort() // Make sure oldest timeslot is first
+        .filter(ts -> isDelayedTimeslotReady(ts, partition, group))
+        .flatMap(timeslot -> this.redisTemplate.opsForSet().remove(key, timeslot)
+            .then(this.redisTemplate.opsForSet().add(key, encodeTimeslotInProgress(timeslot)))
+            .then(Mono.just(timeslot)));
+  }
+
+  private boolean isDelayedTimeslotReady(String timeslot, int partition, String group) {
+    if (isInProgress(timeslot)) {
+      long ts = Long.parseLong(timeslot.split("\\|")[0]);
+      if (ts < (nowEpochSeconds() - appProperties.getIngestStartTime().getSeconds())) {
```

Comment:
This means that we won't need to reset those timeslots to in-progress since they are already in-progress but it should happen so rarely I won't bother to do that check. I will just clutter the code unnecessarily for almost no performance gain.